### PR TITLE
First take at SkyCoord initializer refactor

### PR DESCRIPTION
**Pull request by adrn**
The first thing this PR does is remove an old-standing, deprecated initialization pattern ( [see this line](https://github.com/astropy/astropy/compare/master...adrn:coordinates/skycoord-refactor?expand=1#diff-443cf9a7fca154e62d35de00f15d5c44L245) ), and adjusts tests and docstring to remove any mention of this.

The next thing I've done so far is to re-think the parsing strategy by reworking some of the parsing functions. In it's current state (i.e. on master), the parsing is not modular: frame attributes can be discovered in 3 different functions, and most of the parsing relies on hidden modifications to the main `kwargs` dict. I've tried to expose some of the magic, and group the parsing so that, conceptually, these things happen sequentially:
- Parse arguments to extract a frame class and all frame attributes
- Extract any skycoord-only arguments (e.g., `obstime` not stored on a frame)
- Parse remaining positional/keyword arguments to get the components and prepare the actual coordinate data

The motivation here is to work towards making this more maintainable, but also to (in another PR) allow adding new ways to short-circuit the initialization through a set of classmethods (coming soon). 

_Note: this PR previously contained skeletons for the classmethods, but I've stripped those commits out to simplify_

### CODE REVIEW

**Comment by adrn** [comment]
@eteq Except for the (new, empty) `from_representation()` classmethod, I think the refactor elements of this PR are ready for a first look. (In fact, maybe I should remove the classmethod commit from this and make that a separate PR.) 

The refactoring here is mostly for readability and clarity, but there are some performance gains: this, combined with #7924, lead to factors of 2–7x speed improvements (depending on the exact initialization path) to the SkyCoord initializer.

**Comment by adrn** [comment]
@eteq I've stripped out the new classmethods and left this to be a straight refactor, to hopefully make this a little easier to review.

File: astropy/coordinates/sky_coordinate_parsers.py [hunk 1]
```diff
@@ -0,0 +1,650 @@
+import re
```
Comment (eteq) [inline astropy/coordinates/sky_coordinate_parsers.py]: Might as well add:

```suggestion
# Licensed under a 3-clause BSD style license - see LICENSE.rst

import re
```

File: astropy/coordinates/sky_coordinate_parsers.py [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+from ..units import Unit, IrreducibleUnit
+from .. import units as u
+
+from .baseframe import (BaseCoordinateFrame, frame_transform_graph,
+                        _get_repr_cls, _get_diff_cls,
+                        _normalize_representation_type)
+from .builtin_frames import ICRS
+from .representation import (BaseRepresentation, SphericalRepresentation,
+                             UnitSphericalRepresentation)
+
```
Comment (eteq) [inline astropy/coordinates/sky_coordinate_parsers.py]: Needs a docstring.  Maybe:

```suggestion

"""
This module contains utility functions to make the SkyCoord initializer more modular
and maintainable. No functionality here should be in the public API, but rather used as
part of creating SkyCoord objects. 
"""

```

File: astropy/coordinates/sky_coordinate.py [hunk 3]
```diff
@@ -218,49 +208,69 @@ class or the corresponding string alias.  The frame classes that are built in
...
-        for attr in kwargs:
-            if (attr not in frame_attr_names and
-                attr in frame_transform_graph.frame_attributes):
-                # Setting it will also validate it.
-                setattr(self, attr, kwargs[attr])
+        # If all that is passed in is a frame instance that already has data,
+        # we should bypass all of the parsing and logic below. This is here
+        # to make this the fastest way to create a SkyCoord instance. Many of
+        # the classmethods implemented for performance enhancements will use
+        # this in the initialization path
```
Comment (eteq) [inline astropy/coordinates/sky_coordinate.py]: Extremely minor nitpick:
```suggestion
        # this as the initialization path
```

**Comment by Cadair** [comment]
This is nice, the fact it takes ~700 lines to parse the SkyCoord input boggles the mind!!

**Comment by adrn** [comment]
@Cadair There is still quite a bit of duplication. I want to keep going with the refactor, but am out of time for the feature freeze, so figured this first attempt is still worth it :)

**Pull request merged**